nav2_behavior_tree/test/plugins/decorator/distance_controller_test.cpp

```diff
@@ -0,0 +1,177 @@
+      "node",
+      rclcpp::Node::SharedPtr(transform_handler_));
+    config_->blackboard->set<std::shared_ptr<tf2_ros::Buffer>>(
+      "tf_buffer",
+      transform_handler_->getBuffer());
+    config_->blackboard->set<std::chrono::milliseconds>(
+      "server_timeout",
+      std::chrono::milliseconds(10));
+    config_->blackboard->set<bool>("path_updated", false);
+    config_->blackboard->set<bool>("initial_pose_received", false);
+
+    transform_handler_->activate();
+    transform_handler_->waitForTransform();
+  }
+
+  static void TearDownTestCase()
+  {
+    transform_handler_->deactivate();
+    delete transform_handler_;
+    delete config_;
+    delete dummy_node_;
+    transform_handler_ = nullptr;
+    config_ = nullptr;
+    dummy_node_ = nullptr;
+  }
+
+  void SetUp()
+  {
+    node_ = new nav2_behavior_tree::DistanceController("distance_controller", *config_);
+    node_->setChild(dummy_node_);
+  }
+
+  void TearDown()
+  {
+    node_ = nullptr;
+  }
+
+protected:
+  static nav2_behavior_tree::TransformHandler * transform_handler_;
+  static BT::NodeConfiguration * config_;
+  static nav2_behavior_tree::DistanceController * node_;
+  static nav2_behavior_tree::DummyNode * dummy_node_;
+};
+
+nav2_behavior_tree::TransformHandler * DistanceControllerTestFixture::transform_handler_ = nullptr;
+BT::NodeConfiguration * DistanceControllerTestFixture::config_ = nullptr;
+nav2_behavior_tree::DistanceController * DistanceControllerTestFixture::node_ = nullptr;
+nav2_behavior_tree::DummyNode * DistanceControllerTestFixture::dummy_node_ = nullptr;
+
+TEST_F(DistanceControllerTestFixture, test_initial_status_is_idle)
+{
+  EXPECT_EQ(node_->status(), BT::NodeStatus::IDLE);
+}
+
+TEST_F(DistanceControllerTestFixture, test_failure_when_child_is_idle)
+{
+  dummy_node_->setStatus(BT::NodeStatus::IDLE);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::FAILURE);
+}
+
+TEST_F(DistanceControllerTestFixture, test_failure_when_child_fails)
+{
+  dummy_node_->setStatus(BT::NodeStatus::FAILURE);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::FAILURE);
+}
+
+TEST_F(DistanceControllerTestFixture, test_running_when_child_is_running)
+{
+  dummy_node_->setStatus(BT::NodeStatus::RUNNING);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::RUNNING);
+}
+
+TEST_F(DistanceControllerTestFixture, test_behavior)
+{
+  EXPECT_EQ(node_->status(), BT::NodeStatus::IDLE);
+
+  dummy_node_->setStatus(BT::NodeStatus::SUCCESS);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::SUCCESS);
+  EXPECT_EQ(dummy_node_->status(), BT::NodeStatus::IDLE);
+
+  geometry_msgs::msg::PoseStamped pose;
+  pose.pose.position.x = 0.5;
+  pose.pose.position.y = 0;
+  pose.pose.position.z = 0;
+  pose.pose.orientation.x = 0;
+  pose.pose.orientation.y = 0;
+  pose.pose.orientation.z = 0;
+  pose.pose.orientation.w = 1;
+
+  double traveled = 0;
+
+  transform_handler_->updateRobotPose(pose.pose);
+
+  // Wait for transforms to actually update
+  while (traveled < 0.1) {
```

Comment:
I'm not clear on why this while loop is here. Can't you just `waitForTransform`?

Reply:
What I noticed here was that it takes 4-5 iterations for the pose to actually get updated to the new values. During these iterations, the transforms are available but they are not the updated transforms.

Reply:
even for a waitForTransform at a specific timestamp? If you did wait for transform with a TimePoint(0) to get latest that would happen, but that shouldn't happen if you request for a specific time. 

Overall its a test and I don't really care that much for non-production code, but maybe explain it better in a comment.